File: Algorithms/C++/Easy/28_Implement_strStr().cpp
class Solution {
public:
    int strStr(string haystack, string needle) {
        if (needle.empty()) {
            return 0;
        } else if (haystack.empty()) {
            return -1;
        }

        int haystack_length = haystack.length();
        int needle_length = needle.length();

        for (int i = 0; i < haystack_length - needle_length + 1; i++) {
            int j;
            for (j = 0; j < needle_length && haystack[i + j] == needle[j]; j++) {
                ;
            }

            if (j == needle_length) {
                return i;
            }
        }

        return -1;
    }
};

File: Algorithms/JavaScript/Easy/28_Implement_strStr().js
/**
 * @param {string} haystack
 * @param {string} needle
 * @return {number}
 */
var strStr = function(haystack, needle) {
    if (haystack === "" || needle === "") {
        if (needle === "") {
            return 0
        } else if (haystack === "") {
            return -1
        }
    }

    if (haystack.length < needle.length) {
        return -1
    }

    var index
    for (index = 0; index <= haystack.length - needle.length; index++) {
        var i
        for (i = 0; i < needle.length && haystack[index + i] === needle[i]; i++) ;

        if (i === needle.length) {
            return index
        }
    }

    return -1
};
